Reduce scope of experimental warnings with lookbehind

demerphq opened the pull request:
The only case where we need to warn about variable length lookbehind is
when we are doing postive lookbehind and capturing buffers are involved.

demerphq committed:
regcomp.c: Reduce scope of experimental warnings with lookbehind

The only case where we need to warn about variable length lookbehind is
when capturing buffers are involved.

hvds commented:
First thought here: should it include negative lookbehind too, on the grounds that those captures _can_ come into play within the lookbehind itself? I feel the marginal cost to users of including it would be minimal.

hvds commented on pod/perlre.pod:
```diff
@@ -1652,13 +1652,27 @@ matches a word that follows a tab, without including the tab in C<$&>.
+
+In Perl 5.35.10 the scope of the experimental nature of this construct
+has been reduced, and experimental warnings will only be produced when
+the construct contains capturing parenthesis. The warnings will be
```
s/parenthesis/parentheses/

demerphq replied:
fixed.

hvds reviewed (commented):
The code looks reasonable to me, just one typo.

demerphq commented:
On Sun, 27 Feb 2022, 00:22 Hugo van der Sanden, ***@***.***>
wrote:

> First thought here: should it include negative lookbehind too, on the
> grounds that those captures *can* come into play within the lookbehind
> itself? I feel the marginal cost to users of including it would be minimal
>

At first i thought the same, but I couldn't come up with a single case
where a capture in a negative lookbehind actually captures, and concluded
capturing in nlb (and probably nla) never captures as the construct fails
if it matches, so the capture buffer is cleared. If you can find a case to
the contrary then I would agree that too should be included, not finding a
case quite possibly could be a lack of imagination on my part.

Happy to have secondary confirmation of my analysis.

Thanks for the follow up.

Cheers
Yves

>

hvds commented:
> On Sun, 27 Feb 2022, 00:22 Hugo van der Sanden, ***@***.***>
> wrote:
> > First thought here: should it include negative lookbehind too, on the grounds that those captures *can* come into play within the lookbehind itself? I feel the marginal cost to users of including it would be minimal
> 
> At first i thought the same, but I couldn't come up with a single case where a capture in a negative lookbehind actually captures, and concluded capturing in nlb (and probably nla) never captures as the construct fails if it matches, so the capture buffer is cleared. If you can find a case to the contrary then I would agree that too should be included, not finding a case quite possibly could be a lack of imagination on my part.

By "within the lookbehind itself" I meant cases such as `/(?<!(.)\1)/`. Trying it with a handy 5.34 now, I find that is disallowed ("Lookbehind longer than 255 not implemented"). I haven't checked, but I'd guess that because the length of `\1` isn't known it gets treated as infinite.

My inclination would be to include it with the positive case anyway: if it is never useful, the marginal cost drops to zero, while the value of the wiggle room it gives us is slim but non-zero. It would also make both the implementation and the explanation a bit simpler.

demerphq commented:
On Sun, 27 Feb 2022 at 02:20, Hugo van der Sanden ***@***.***>
wrote:

> On Sun, 27 Feb 2022, 00:22 Hugo van der Sanden, *@*.***>
> wrote:
>
> First thought here: should it include negative lookbehind too, on the
> grounds that those captures *can* come into play within the lookbehind
> itself? I feel the marginal cost to users of including it would be minimal
>
> At first i thought the same, but I couldn't come up with a single case
> where a capture in a negative lookbehind actually captures, and concluded
> capturing in nlb (and probably nla) never captures as the construct fails
> if it matches, so the capture buffer is cleared. If you can find a case to
> the contrary then I would agree that too should be included, not finding a
> case quite possibly could be a lack of imagination on my part.
>
> By "within the lookbehind itself" I meant cases such as /(?<!(.)\1)/.
> Trying it with a handy 5.34 now, I find that is disallowed ("Lookbehind
> longer than 255 not implemented"). I haven't checked, but I'd guess that
> because the length of \1 isn't known it gets treated as infinite.
>
Oh! Good play! I had not thought of /that/. Make sense now!

If we allowed negative lookbehind with captures without the experimental
flag and then someone tweaked this logic then we would be back in undefined
territory, so I agree that we should include negative.


> My inclination would be to include it with the positive case anyway: if it
> is never useful, the marginal cost drops to zero, while the value of the
> wiggle room it gives us is slim but non-zero. It would also make both the
> implementation and the explanation a bit simpler.
>
Your example above seals the deal for me, I'll adjust the patch
accordingly. Nice one.

Cheers,
yves
-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

hvds commented:
I mean something like this:
```
# named capture correctly triggers warning: ok
% ./miniperl -we 'qr{ (?=(?<c>x)) | (?<=(?<d>y)|zz) }x'
Variable length positive lookbehind with capturing is experimental in regex; marked by <-- HERE \
  in m/ (?=(?<c>x)) | (?<=(?<d>y)|zz)  <-- HERE / at -e line 1.
% 

# .. even if reused: ok
% ./miniperl -we 'qr{ (?=(?<c>x)) | (?<=(?<c>y)|zz) }x'
Variable length positive lookbehind with capturing is experimental in regex; marked by <-- HERE \
  in m/ (?=(?<c>x)) | (?<=(?<c>y)|zz)  <-- HERE / at -e line 1.
% 

# reuse of numbered capture via (?| ...) not ok
% ./miniperl -we 'qr{ (?| (?=(x)) | (?<=(y)|zz) ) }x'
% 
```

That last case needs looking at: it should warn too.

demerphq commented:
On Sun, 27 Feb 2022 at 14:06, Hugo van der Sanden ***@***.***>
wrote:

> I mean something like this:
>
> # named capture correctly triggers warning: ok
> % ./miniperl -we 'qr{ (?=(?<c>x)) | (?<=(?<d>y)|zz) }x'
> Variable length positive lookbehind with capturing is experimental in regex; marked by <-- HERE \
>   in m/ (?=(?<c>x)) | (?<=(?<d>y)|zz)  <-- HERE / at -e line 1.
> %
>
> # .. even if reused: ok
> % ./miniperl -we 'qr{ (?=(?<c>x)) | (?<=(?<c>y)|zz) }x'
> Variable length positive lookbehind with capturing is experimental in regex; marked by <-- HERE \
>   in m/ (?=(?<c>x)) | (?<=(?<c>y)|zz)  <-- HERE / at -e line 1.
> %
>
> # reuse of numbered capture via (?| ...) not ok
> % ./miniperl -we 'qr{ (?| (?=(x)) | (?<=(y)|zz) ) }x'
> %
>
> That last case needs looking at: it should warn too
>
Gnash. This is like the 12th time that stupid thing has come back to bite
me. If we had had experimental back in the day it have been marked
experimental, and it would still be marked experimental it has caused so
much trouble.

Thanks for bringing this to my attention. That *deeply* sucks.

Cheers,
Yves

-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

hvds commented:
> That *deeply* sucks.

Well, while not ideal, it would not be disastrous if that's the only construct we'd fail to notice: it's a deeply weird thing to do.

Another case to rule in or out is GOSUB; I have a half-memory we might wrap an ersatz capture around it as part of the implementation, but I haven't verified that.

khwilliamson commented on t/porting/known_pod_issues.dat:
```diff
@@ -389,6 +389,7 @@ pod/perlandroid.pod	Verbatim line length including indents exceeds 78 by	3
 pod/perlbook.pod	Verbatim line length including indents exceeds 78 by	1
 pod/perldebguts.pod	Verbatim line length including indents exceeds 78 by	24
 pod/perldebtut.pod	Verbatim line length including indents exceeds 78 by	2
+pod/perldiag.pod	Apparent broken link	2
```
There's something wrong here that should be fixed.

demerphq replied:
Fixed now.

khwilliamson reviewed (approved):
Otherwise LGTM

demerphq commented:
On Mon, 28 Feb 2022, 05:34 Karl Williamson, ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In t/porting/known_pod_issues.dat
> <https://github.com/Perl/perl5/pull/19455#discussion_r815494897>:
>
> > @@ -389,6 +389,7 @@ pod/perlandroid.pod	Verbatim line length including indents exceeds 78 by	3
>  pod/perlbook.pod	Verbatim line length including indents exceeds 78 by	1
>  pod/perldebguts.pod	Verbatim line length including indents exceeds 78 by	24
>  pod/perldebtut.pod	Verbatim line length including indents exceeds 78 by	2
> +pod/perldiag.pod	Apparent broken link	2
>
> There's something wrong here that should be fixed
>

Podchecker doesn't seem to grok X<> style index entries, I tried several
things, and none of them worked and I eventually grew tired of wrestling
with it and used the --regen escape hatch. Promise i tried pretty hard to
get it to pass without using that! I'd truly love to know what I could have
done differently if you have advice. The cause is the two L<> references to
the docs for positive and negative lookbehind I added to the perldiag
entries.

Yves

>

demerphq commented:
On Mon, 28 Feb 2022, 00:24 Hugo van der Sanden, ***@***.***>
wrote:

> That *deeply* sucks.
>
> Well, while not ideal, it would not be disastrous if that's the only
> construct we'd fail to notice: it's a deeply weird thing to do.
>
I agree. But I think I can probably work out a solution, although a priori
I wonder how invasive it might be.

Another case to rule in or out is GOSUB; I have a half-memory we might wrap
> an ersatz capture around it as part of the implementation, but I haven't
> verified that.
>

I imagine gosub out of one of these constructs is forbidden, but I'll check
and if necessary take care if it, gosub in should be fine.

Thanks for the pointers, I'll check up on it.

Yves

>

hvds commented:
> The cause is the two L<> references to the docs for positive and negative lookbehind I added to the perldiag entries.

Why have two diagnostics rather than one? When I suggested earlier that including negative lookbehind too would "also make both the implementation and the explanation a bit simpler", it was in part with the assumption that things like warning messages would simply lose "positive" from their text.

demerphq commented:
On Mon, 28 Feb 2022, 21:15 Hugo van der Sanden, ***@***.***>
wrote:

> The cause is the two L<> references to the docs for positive and negative
> lookbehind I added to the perldiag entries.
>
> Why have two diagnostics rather than one? When I suggested earlier that
> including negative lookbehind too would "also make both the implementation
> and the explanation a bit simpler", it was in part with the assumption that
> things like warning messages would simply lose "positive" from their text
>
Because they are different constructs with different issues, and rather
different example cases and explanations. As a user I want more detailed
tailored errors not generic ones. Anyway, lookbehind isn't documented as a
single thing with two flavors but rather as a set of two different
constructs each of which can be spelled three different ways. Which would I
link to? Both? Rewrite the docs totally? Why not DTRT instead?

I don't see any real win with this simpler implementation you propose, it's
a neglible drop in a large ocean. I'd rather have the tiny added
complexity, which is really nothing at all in the grand scheme of things,
if it gives a more useful user experience. We should be optimizing for our
users not us as core devs with this kind of thing.

Also it wouldn't make a difference to the subject of this subthread, there
would still be a bogus report of a broken link, or two, podchecker is just
broken in how it handles X<> index entries and L<> to them. The code, docs
and diags are written, if you want to spend energy on this subject I think
it would better placed fixing the pod checker if I may be so bold to say,
it would be of far greater benefit to the project as a whole than undoing
the work I already have done.

Cheers
Yves

hvds commented:
> > > The cause is the two L<> references to the docs for positive and negative lookbehind I added to the perldiag entries.
> >
> > Why have two diagnostics rather than one?
> 
> [...] it wouldn't make a difference to the subject of this subthread, there would still be a bogus report of a broken link, or two, podchecker is just broken in how it handles X<> index entries and L<> to them.

Ah, my apologies; I interpreted your original comment above as implying that the problem occurred _because_ there were two entries.

> if you want to spend energy on this subject [...]

I was just trying to help, no offence or sapping of energy was intended.

khwilliamson commented:
On 2/28/22 06:36, Yves Orton wrote:
> On Mon, 28 Feb 2022, 21:15 Hugo van der Sanden, ***@***.***>
> wrote:
> 
>  > The cause is the two L<> references to the docs for positive and negative
>  > lookbehind I added to the perldiag entries.
>  >
>  > Why have two diagnostics rather than one? When I suggested earlier that
>  > including negative lookbehind too would "also make both the 
> implementation
>  > and the explanation a bit simpler", it was in part with the 
> assumption that
>  > things like warning messages would simply lose "positive" from their text
>  >
> Because they are different constructs with different issues, and rather
> different example cases and explanations. As a user I want more detailed
> tailored errors not generic ones. Anyway, lookbehind isn't documented as a
> single thing with two flavors but rather as a set of two different
> constructs each of which can be spelled three different ways. Which would I
> link to? Both? Rewrite the docs totally? Why not DTRT instead?
> 
> I don't see any real win with this simpler implementation you propose, it's
> a neglible drop in a large ocean. I'd rather have the tiny added
> complexity, which is really nothing at all in the grand scheme of things,
> if it gives a more useful user experience. We should be optimizing for our
> users not us as core devs with this kind of thing.
> 
> Also it wouldn't make a difference to the subject of this subthread, there
> would still be a bogus report of a broken link, or two, podchecker is just
> broken in how it handles X<> index entries and L<> to them. The code, docs
> and diags are written, if you want to spend energy on this subject I think
> it would better placed fixing the pod checker if I may be so bold to say,
> it would be of far greater benefit to the project as a whole than undoing
> the work I already have done.
> 

It appears you think that X<> entries are what gets linked to.  That is 
incorrect, and explains why you couldn't get it to work.

X<> is used for indexing, not linking.  As the docs say:

X<topic name>" -- an index entry
          This is ignored by most formatters, but some may use it for 
building indexes. It always renders as empty-string.

Maybe buildtoc uses them; I've never looked into it.  That would be the 
only place I can think of that uses them in the Perl core.

L<> is used for linking. The target must be a "section" in the linked-to 
page.  A "section" is introduced by =head or =item.

perlpodspec says

Note that section names might contain markup. I.e., if a section
          starts with:

            =head2 About the C<-M> Operator

          or with:

            =item About the C<-M> Operator

          then a link to it would look like this:

          L<somedoc/About the C<-M> Operator>

The podchecker is correctly doing its job here.

If it were me, I would make the link
   L<perlre/Lookaround Assertions>

and make them start reading/skimming there.

demerphq commented:
On Mon, 28 Feb 2022 at 18:18, Hugo van der Sanden ***@***.***>
wrote:

> The cause is the two L<> references to the docs for positive and negative
> lookbehind I added to the perldiag entries.
>
> Why have two diagnostics rather than one?
>
> [...] it wouldn't make a difference to the subject of this subthread,
> there would still be a bogus report of a broken link, or two, podchecker is
> just broken in how it handles X<> index entries and L<> to them.
>
> Ah, my apologies; I interpreted your original comment above as implying
> that the problem occurred *because* there were two entries.
>
No, sorry if I wasnt clear.


> if you want to spend energy on this subject [...]
>
> I was just trying to help, no offence or sapping of energy was intended.
>

I just meant that if you were thinking of reworking all the docs it might
be better to solve the other problem. I didnt take offense, sorry if it
seemed like it. I just dont think reworking the docs so we could have one
section on lookbehind is worth your or my time. I do think that improving
how podchecker is worth your or my time. It's on my list, but I'm in the
middle of other tasks already.

Yves

demerphq commented:
On Tue, 1 Mar 2022 at 05:15, Karl Williamson ***@***.***>
wrote:

> On 2/28/22 06:36, Yves Orton wrote:
> > On Mon, 28 Feb 2022, 21:15 Hugo van der Sanden, ***@***.***>
> > wrote:
> >
> > > The cause is the two L<> references to the docs for positive and
> negative
> > > lookbehind I added to the perldiag entries.
> > >
> > > Why have two diagnostics rather than one? When I suggested earlier that
> > > including negative lookbehind too would "also make both the
> > implementation
> > > and the explanation a bit simpler", it was in part with the
> > assumption that
> > > things like warning messages would simply lose "positive" from their
> text
> > >
> > Because they are different constructs with different issues, and rather
> > different example cases and explanations. As a user I want more detailed
> > tailored errors not generic ones. Anyway, lookbehind isn't documented as
> a
> > single thing with two flavors but rather as a set of two different
> > constructs each of which can be spelled three different ways. Which
> would I
> > link to? Both? Rewrite the docs totally? Why not DTRT instead?
> >
> > I don't see any real win with this simpler implementation you propose,
> it's
> > a neglible drop in a large ocean. I'd rather have the tiny added
> > complexity, which is really nothing at all in the grand scheme of things,
> > if it gives a more useful user experience. We should be optimizing for
> our
> > users not us as core devs with this kind of thing.
> >
> > Also it wouldn't make a difference to the subject of this subthread,
> there
> > would still be a bogus report of a broken link, or two, podchecker is
> just
> > broken in how it handles X<> index entries and L<> to them. The code,
> docs
> > and diags are written, if you want to spend energy on this subject I
> think
> > it would better placed fixing the pod checker if I may be so bold to say,
> > it would be of far greater benefit to the project as a whole than undoing
> > the work I already have done.
> >
>
> It appears you think that X<> entries are what gets linked to. That is
> incorrect, and explains why you couldn't get it to work.
>
> X<> is used for indexing, not linking. As the docs say:
>

Yes, I assumed that something that is marked as an index entry can be
linked to. I guess you are saying it doesn't work like that. I am not sure
how I could have known that however.


>
> X<topic name>" -- an index entry
> This is ignored by most formatters, but some may use it for
> building indexes. It always renders as empty-string.
>
> Maybe buildtoc uses them; I've never looked into it. That would be the
> only place I can think of that uses them in the Perl core.
>
> L<> is used for linking. The target must be a "section" in the linked-to
> page. A "section" is introduced by =head or =item.


I had formed the impression that X<> entries were  a way to give
alternative names to the sections. I now understand I was mistaken. Thanks
for explaining that, was driving me bonkers.


>
> perlpodspec says
>
> Note that section names might contain markup. I.e., if a section
> starts with:
>
> =head2 About the C<-M> Operator
>
> or with:
>
> =item About the C<-M> Operator
>
> then a link to it would look like this:
>
> L<somedoc/About the C<-M> Operator>
>
> The podchecker is correctly doing its job here.
>
> If it were me, I would make the link
> L<perlre/Lookaround Assertions>
>
> and make them start reading/skimming there.
>

Thanks for the guidance, ill take a look. I would prefer to link the
correct section instead.

cheers.
Yves

demerphq commented:
On Tue, 1 Mar 2022 at 08:20, demerphq ***@***.***> wrote:
> Thanks for the guidance, ill take a look. I would prefer to link the correct section instead.

Fixed and repushed.

Cheers,
Yves
-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

demerphq committed:
regcomp.c: S_study_chunk() clean up args and docs

Put one argument per line, add a comment per argument,
and remove docs for long removed argument "recursed".

demerphq committed:
regcomp.c: rename "fake" argument in main entry points to study_chunk()

There are way too many uses of "fake" as a variable in this code, this
renames two of them to improve clarity. Both are used for the "deltap"
argument to study_chunk() which is not used in the top level calls.

demerphq committed:
regcomp.c: rename "fake" where it used in recursive study_chunk() calls

All instances of the fake variable are used inside of data_fake.last_closep
structures, so rename them from "fake" to "fake_last_close" so it is obvious
what they are used for. Note the 'p' in last_closep means "pointer", eg
it means "pointer to last_close", so the "fake" version is called
"fake_last_close" as well.

demerphq committed:
regcomp.c: track last close *op*, not just its index in study_chunk()

In the study_data_t struct the member last_closep is a pointer to a
SSize_t which is used to store the arg (which buffer slot it represents,
eg, $1 or $3 at the perl level) from the last close regop we encountered
during the scan process.

The arg might be duplicated between different opcodes in the face of
branch-reset /(?| (...) | (...) )/ where both capture buffers map to the
same perl variable (eg, $1).

This patches adds a new, very similar member, regnode **last_close_opp,
which tracks the close op itself instead of its arg.

In a future commit we could replace last_closep entirely, and extract
the comparable value from the regnode pointed to by last_close_opp when
needed, but for this patch I am sticking with tracking both at the same
time as it is simpler and safer.

This enables us to detect if someone has used parens inside of a
parenthesized construct reliably in the fact of branch reset.

This is in preparation for dealing with experimental variable length
warnings properly, which will come in a subsequent patch. It also should
help long run to fixing other bugs related to branch-reset.

NB: Really a bunch of this should turn into some kind of macro to
    simplify the code and ensure that this is done reliably without a
    lot of repeated similar but not quite identical logic. But that too
    is not appropriate for this PR.

demerphq committed:
regcomp.c: in study_chunk() properly pass down SCF_TRIE_DOING_RESTUDY flag

This is necessary so that warnings generated during study_chunk() do not
get emitted twice. At the same time fix some "int" flags parameters that
should be U32.

demerphq commented:
On Sun, 27 Feb 2022 at 14:10, demerphq ***@***.***> wrote:
>
> On Sun, 27 Feb 2022 at 14:06, Hugo van der Sanden ***@***.***> wrote:
>>
>> I mean something like this:
>>
>> # named capture correctly triggers warning: ok
>> % ./miniperl -we 'qr{ (?=(?<c>x)) | (?<=(?<d>y)|zz) }x'
>> Variable length positive lookbehind with capturing is experimental in regex; marked by <-- HERE \
>>   in m/ (?=(?<c>x)) | (?<=(?<d>y)|zz)  <-- HERE / at -e line 1.
>> %
>>
>> # .. even if reused: ok
>> % ./miniperl -we 'qr{ (?=(?<c>x)) | (?<=(?<c>y)|zz) }x'
>> Variable length positive lookbehind with capturing is experimental in regex; marked by <-- HERE \
>>   in m/ (?=(?<c>x)) | (?<=(?<c>y)|zz)  <-- HERE / at -e line 1.
>> %

So named captures are just captures, so no need for special tests there.

>>
>> # reuse of numbered capture via (?| ...) not ok
>> % ./miniperl -we 'qr{ (?| (?=(x)) | (?<=(y)|zz) ) }x'
>> %
>>
>> That last case needs looking at: it should warn too
>
> Gnash. This is like the 12th time that stupid thing has come back to bite me. If we had had experimental back in the day it have been marked experimental, and it would still be marked experimental it has caused so much trouble.
>
> Thanks for bringing this to my attention. That *deeply* sucks.

I pushed fixes for this today.  And a bunch of tests for the
experimental warnings that were missing. So this should be fixed now.
Also fixed a couple of other things that were in the way to getting
this fixed.

cheers,
Yves
Ps: Thanks for the reminder. There are a bunch of other branch-reset
related bugs and this reminded me to address them so Ill be doing more
work on this subject in the next days.

hvds commented:
> I pushed fixes for this today. And a bunch of tests for the experimental warnings that were missing. So this should be fixed now. Also fixed a couple of other things that were in the way to getting this fixed.

LGTM; I commented with a couple of minor opportunities for improvement, but I'm happy enough with it as is.

demerphq commented:
I totally reworked the patch sequence so that all of the ancillary changes are self contained, and the changes to actually reduce the scope of the warning are in a single self contained last patch. I also added comments as suggested by hugo. I tested each commit one at a time with 'make test_reonly and they all pass. About the only thing I haven't done is spell checked the new comments and commit messages. I will do that later and maybe repush if necessary. I will also prepare a new "remove experimental status" patch based on top of this. I expect to merge this today.

hvds commented:
> I assume you've thought about this, but how do we know that this op will still be around by the time we come to look at it? Is it just that we happen to know a close op is never modified by in-place optimizations? I don't know, for example, whether the TRIE replacement can ever move or replace a capture.
> 
> If there's an invariant we're relying on here, it would be useful to make sure it's noted somewhere.

Not sure why github didn't capture the line being commented on, it should probably have been in 84950019128:
```
+                *(data->last_close_opp) = scan;
```

demerphq commented:
On Thu, 3 Mar 2022 at 12:59, Hugo van der Sanden ***@***.***>
wrote:

> I assume you've thought about this, but how do we know that this op will
> still be around by the time we come to look at it? Is it just that we
> happen to know a close op is never modified by in-place optimizations? I
> don't know, for example, whether the TRIE replacement can ever move or
> replace a capture.
>
> If there's an invariant we're relying on here, it would be useful to make
> sure it's noted somewhere.
>
> Not sure why github didn't capture the line being commented on, it should
> probably have been in 8495001
> <https://github.com/Perl/perl5/commit/8495001912886fcd4f9ecece064aae6f78bcbebd>
> :
>
> +                *(data->last_close_opp) = scan;
>
> Yeah, it's a pointer to a pointer to a regnode. The first pointer is
"safe" and we dereference it to look at the scan pointer, which as you
pointed out is not.

Yves

-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

demerphq commented:
Ok, merging, Ill do some follow ups after. This really cleans up the code for the stuff im working on.
